filebeat/crawler/prospector_log.go

```diff
@@ -38,81 +39,75 @@ func NewProspectorLog(config cfg.ProspectorConfig, channel chan *input.FileEvent
 	return prospectorer, nil
 }
 
-func (prospector ProspectorLog) Init() {
-
-	// Seed last scan time
-	prospector.lastscan = time.Now()
-
-	logp.Debug("prospector", "exclude_files: %s", prospector.config.ExcludeFiles)
-
-	// Now let's do one quick scan to pick up new files
-	for _, path := range prospector.config.Paths {
-		prospector.scan(path, prospector.channel)
-	}
+func (p ProspectorLog) Init() {
+	logp.Debug("prospector", "exclude_files: %s", p.config.ExcludeFiles)
+	p.scan()
 }
 
-func (prospector ProspectorLog) Run(spoolChan chan *input.FileEvent) {
+func (p ProspectorLog) Run() {
 
-	newlastscan := time.Now()
+	logp.Debug("prospector", "Start next scan")
+	p.scan()
 
-	for _, path := range prospector.config.Paths {
-		prospector.scan(path, spoolChan)
+	// Clear out files that disappeared and we've stopped harvesting
+	for file, lastinfo := range p.prospectorList {
+		if lastinfo.Finished() && lastinfo.LastIteration < p.iteration {
+			delete(p.prospectorList, file)
+		}
 	}
 
-	prospector.lastscan = newlastscan
+	p.iteration++ // Overflow is allowed
 
 	// Defer next scan for the defined scanFrequency
-	time.Sleep(prospector.config.ScanFrequencyDuration)
-	logp.Debug("prospector", "Start next scan")
+	time.Sleep(p.config.ScanFrequencyDuration)
```

Comment:
The sleep function. Something like util.sleep(...) and it would also have the exit channel from the beat object (somehow ...).